File: Problems/Algorithms/587. Erect the Fence/ErectFence.java
class Solution {
    public int orientation(int[] p, int[] q, int[] r) {
        return (q[0] - p[0]) * (r[1] - q[1]) - (q[1] - p[1]) * (r[0] - q[0]);
    }
    
    public int[][] outerTrees(int[][] trees) {
        int n = trees.length;
        if (n < 3) {
            return trees;
        }

        Arrays.sort(trees, new Comparator<int[]> () {
            public int compare(int[] p, int[] q) {
                return q[0] - p[0] == 0 ? q[1] - p[1] : q[0] - p[0];
            }
        });
        
        Stack<int[]> convexHull = new Stack<>();
        for (int i = 0; i < n; i++) {
            int[] curr = trees[i];
            while (convexHull.size() >= 2 && orientation(convexHull.get(convexHull.size()-2), convexHull.get(convexHull.size()-1), curr) < 0) {
                convexHull.pop();
            }
            convexHull.push(curr);
        }
        for (int i = n-2; i >= 0; i--) {
            int[] curr = trees[i];
            while (convexHull.size() >= 2 && orientation(convexHull.get(convexHull.size()-2), convexHull.get(convexHull.size()-1), curr) < 0) {
                convexHull.pop();
            }
            convexHull.push(curr);
        }

        HashSet<int[]> set = new HashSet<>(convexHull);
        return set.toArray(new int[set.size()][]);
    }
}

File: Problems/Algorithms/587. Erect the Fence/erect_fence.py
#!/usr/bin/env python3
# coding: utf-8

from typing import List


class Solution:
    def outerTrees(self, trees: List[List[int]]) -> List[List[int]]:
        def cross(o, a, b):
            return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])
        
        if len(trees) <= 1:
            return []

        trees.sort(key=lambda x: (x[0], x[1]))

        lower = []
        for p in trees:
            while len(lower) >= 2 and cross(lower[-2], lower[-1], p) < 0:
                lower.pop()
            lower.append(p)
        
        upper = []
        for p in reversed(trees):
            while len(upper) >= 2 and cross(upper[-2], upper[-1], p) < 0:
                upper.pop()
            upper.append(p)
        
        res = []
        for tree in lower[:-1] + upper[:-1]:
            if tree not in res:
                res.append(tree)
        return res
